From 9590ce95f1646e2597df8c639c760702252d4fe4 Mon Sep 17 00:00:00 2001
From: Ankit Agrawal <ankit.agarwal@intel.com>
Date: Tue, 19 Dec 2023 10:06:46 +0530
Subject: [PATCH] Fixed for Media crash during Monkey Test Run.

Observed following crash
java.lang.NullPointerException Attempt to invoke virtual method
'java.lang.Object android.car.Car.getCarManager(java.lang.String)' on a
null object reference.

Creating Car object based on car service listener. Once car service is
present, it will get connected and get the car object. Once get the car
object, then it can get the media manager successfully.

Tests:
Run monkey test and observe, test is pass.

Tracked-On: OAM-114265
Signed-off-by: Ankit Agrawal <ankit.agarwal@intel.com>
---
 .../common/source/MediaSourceViewModel.java   | 19 +++++++++++--------
 1 file changed, 11 insertions(+), 8 deletions(-)

diff --git a/car-media-common/src/com/android/car/media/common/source/MediaSourceViewModel.java b/car-media-common/src/com/android/car/media/common/source/MediaSourceViewModel.java
index 00fb22b57..0d15e74c9 100644
--- a/car-media-common/src/com/android/car/media/common/source/MediaSourceViewModel.java
+++ b/car-media-common/src/com/android/car/media/common/source/MediaSourceViewModel.java
@@ -45,7 +45,7 @@ public class MediaSourceViewModel extends AndroidViewModel {
     private static final String TAG = "MediaSourceViewModel";
 
     private static MediaSourceViewModel[] sInstances = new MediaSourceViewModel[2];
-    private final Car mCar;
+    private Car mCar;
     private CarMediaManager mCarMediaManager;
 
     // Primary media source.
@@ -125,12 +125,14 @@ public class MediaSourceViewModel extends AndroidViewModel {
 
         mMode = mode;
         mInputFactory = inputFactory;
-        mCar = inputFactory.getCarApi((car, ready) -> {
+        inputFactory.getCarApi((car, ready) -> {
+            mCar = ready ? car : null;
             if (!ready) {
                 Log.d(TAG, "Disconnected from Car Service");
                 return;
             }
             Log.d(TAG, "Connected to Car Service");
+            mCarMediaManager = mInputFactory.getCarMediaManager(car);
         });
 
         mBrowserConnector = inputFactory.createMediaBrowserConnector(application, mBrowserCallback);
@@ -139,23 +141,22 @@ public class MediaSourceViewModel extends AndroidViewModel {
         mMediaSourceListener = componentName -> mHandler.post(
                 () -> updateModelState(mInputFactory.getMediaSource(componentName)));
 
-        try {
-            mCarMediaManager = mInputFactory.getCarMediaManager(mCar);
+        if (mCarMediaManager != null) {
             mCarMediaManager.addMediaSourceListener(mMediaSourceListener, mode);
             MediaSource src = mInputFactory.getMediaSource(mCarMediaManager.getMediaSource(mode));
             if (Log.isLoggable(TAG, Log.INFO)) {
                 Log.i(TAG, "Initializing with " + src);
             }
             updateModelState(src);
-        } catch (CarNotConnectedException e) {
-            Log.e(TAG, "Car not connected", e);
         }
     }
 
     @Override
     protected void onCleared() {
         super.onCleared();
-        mCar.disconnect();
+        if (mCar != null) {
+            mCar.disconnect();
+        }
     }
 
     @VisibleForTesting
@@ -180,7 +181,9 @@ public class MediaSourceViewModel extends AndroidViewModel {
         } else if (Log.isLoggable(TAG, Log.WARN)) {
             Log.w(TAG, "Inconsistent media source mode " + mode + " mMode: " + mMode);
         }
-        mCarMediaManager.setMediaSource(mediaSource.getBrowseServiceComponentName(), mode);
+        if (mCarMediaManager != null) {
+            mCarMediaManager.setMediaSource(mediaSource.getBrowseServiceComponentName(), mode);
+        }
     }
 
     /**
-- 
2.17.1

